### DISCUSSION TITLE
[need help]Why generate many import expression in dist/index.mjs ?

### DISCUSSION THREAD

**Discussion by vaynevayne** [rollup/rollup#5459] [Q&A] [answered]
<img width="1261" alt="image" src="https://github.com/rollup/rollup/assets/124666577/35e928b2-b31a-4f6f-ab80-69be5483fd48">

```jsx
import { defineConfig } from "rollup";
import resolvePlugin from "@rollup/plugin-node-resolve";
import commonjs from "@rollup/plugin-commonjs";
import { cwd } from "process";
import { globSync } from "glob";
import path from "node:path";
import { readFileSync } from "fs";
import esbuild from "rollup-plugin-esbuild";
import image from "@rollup/plugin-image";

const packageJson = JSON.parse(
  readFileSync(path.resolve(cwd(), "./package.json"))
);

const isProd = process.env.NODE_ENV === "production";

const defaultPlugins = [
  resolvePlugin(),
  image(),
  commonjs({
    include: /node_modules/,
  }),
];

const onwarn = (warning, rollupWarn) => {
  rollupWarn(warning);
};

const deps = [
  ...Object.keys(packageJson.dependencies ?? {}),
  ...Object.keys(packageJson.peerDependencies ?? {}),
];

const external = new RegExp(`^(${deps.join("|")})`);

const emsConfig = defineConfig({
  input: Object.fromEntries(
    globSync("src/**/*.{ts,tsx}").map((file) => {

      return [
  
        path.relative(
          "src",
          file.slice(0, file.length - path.extname(file).length)
        ),
    
        path.resolve(cwd(), file),
      ];
    })
  ),
  output: {
    format: "es",
    exports: "named",
    entryFileNames: "[name].mjs",
    dir: path.resolve(cwd(), "dist/esm"),
  },
  onwarn,
  external,
  plugins: [
    ...defaultPlugins,
    esbuild({
      sourceMap: true,
      tsconfig: path.resolve(cwd(), "tsconfig.json"),
      // minify: process.env.NODE_ENV === "production",
    }),
  ],
});
```

**Comment by lukastaegert** [comment, marked as answer]
https://rollupjs.org/faqs/#why-do-additional-imports-turn-up-in-my-entry-chunks-when-code-splitting

**Reply by vaynevayne** [reply]
Thank you. According to this matter, I think of a point of view. By default, it will be loaded in advance,
In fact, it will lead to treeshaink failure. How can the article avoid talking about it?

**Reply by lukastaegert** [reply]
It will not cause tree-shaking issues for Rollup. Not sure why it would cause issues for other bundlers.

**Reply by vaynevayne** [reply]
```js
import 'lodash'
```
When the project is packaged, it is found that this file has an import' loadsh' statement, then it will download lodash and package it into the file, resulting in two problems:
1. The size of the package becomes larger, even if lodash is not used in the back,
2. Assuming that only one function of lodash is used, it will also cause the entire lodash to be downloaded.
Is my understanding right?

**Reply by lukastaegert** [reply]
As lodash is a static dependency of one of your files, it will be included no matter whether the top-level file imports it or whether a file imported by an imported file imports it. Rollup's logic does not change that, it is your code that is importing it on the critical path, Rollup just exposes it.
If you do not want lodash to be imported unconditionally, you need to put the file that relies on it behind a dynamic import.
